### PR TITLE
Life cycle of Showback_Bucket

### PR DESCRIPTION
**Pull request by aljesusg**
CU

-  STATE SHOULD GO  open -> processing -> close

### CODE REVIEW

File: app/models/showback_bucket.rb [hunk 1]
```diff
@@ -1,8 +1,36 @@
...
+
+  def check_bucket_state
+    case state_was
+      when "OPEN"       then  raise _("Bucket can't pass to CLOSE after OPEN")      unless state != "CLOSE"
+      when "PROCESSING"
+        raise _("Bucket can't pass to OPEN after PROCESSING") unless state != "OPEN"
+        ShowbackBucket.create(:name        => self.name,
+                              :description => self.description,
+                              :resource    => self.resource,
+                              :start_time  => (self.start_time + 1.months).beginning_of_month,
```
Comment (sergio-ocon) [inline app/models/showback_bucket.rb]: Wouldn't it be?

if self.end_time == self.end_of_month do
  new_start_time = self.next_month.beginning_of_month
else 
 new_start_time = self.end_time

:start_time => new_start_time
:end_time => new.start_time.end_of_month

File: app/models/showback_bucket.rb [hunk 2]
```diff
@@ -1,8 +1,36 @@
...
+  validates :description,           :presence => true
+  validates :resource,              :presence => true
+  validates :start_time, :end_time, :presence => true
+  validates :state,                 :presence => true, :inclusion => { :in => %w(OPEN PROCESSING CLOSE) }
+
+  #End_time should be after start_time.
+  validate  :start_time_before_end_time
+
+  def start_time_before_end_time
+    errors.add(:start_time, "Start time should be before end time") unless end_time.to_i > start_time.to_i
```
Comment (sergio-ocon) [inline app/models/showback_bucket.rb]: _() to allow i18n

File: db/migrate/20170508131106_add_state_showback_bucket.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+class AddStateShowbackBucket < ActiveRecord::Migration[5.0]
+  def up
+    change_table :showback_buckets do |t|
+      t.timestamp   :start_time
```
Comment (sergio-ocon) [inline db/migrate/20170508131106_add_state_showback_bucket.rb]: Time.all_month returns an interval, can we use an interval in queries?

**Comment by aljesusg** [comment]
Should I make here the log feature or in other PR @sergio-ocon ?

File: app/models/showback_bucket.rb (hunk 1 above)
Comment (aljesusg) [inline app/models/showback_bucket.rb]: Rdy

File: app/models/showback_bucket.rb (hunk 2 above)
Comment (aljesusg) [inline app/models/showback_bucket.rb]: Rdy

File: .rubocop_local.yml [hunk 4]
```diff
@@ -18,6 +18,7 @@ GlobalVars:
   - $scvmm_log
   - $vim_log
   - $websocket_log
+  - $miq_showback
```
Comment (sergio-ocon) [inline .rubocop_local.yml]: Let's add .rubocop_local.yml to the ignored files

File: app/models/showback_bucket.rb [hunk 5]
```diff
@@ -1,8 +1,42 @@
...
+  def check_bucket_state
+    case state_was
+      when "OPEN"  then raise _("Bucket can't pass to CLOSE after OPEN")      unless state != "CLOSE"
+      when "PROCESSING"
+        raise _("Bucket can't pass to OPEN after PROCESSING") unless state != "OPEN"
+        s_time = (self.start_time + 1.months).beginning_of_month
+        e_time = (self.end_time + 1.months).end_of_month
+        if self.end_time != self.start_time.end_of_month
+            s_time = self.end_time
+            e_time = s_time.end_of_month
```
Comment (sergio-ocon) [inline app/models/showback_bucket.rb]: e_time can always be s_time.end_of_month

File: app/models/showback_bucket.rb [hunk 6]
```diff
@@ -1,8 +1,42 @@
...
+  def start_time_before_end_time
+    errors.add(:start_time, _("Start time should be before end time")) unless end_time.to_i > start_time.to_i
+  end
+
+  def check_bucket_state
+    case state_was
+      when "OPEN"  then raise _("Bucket can't pass to CLOSE after OPEN")      unless state != "CLOSE"
+      when "PROCESSING"
+        raise _("Bucket can't pass to OPEN after PROCESSING") unless state != "OPEN"
+        s_time = (self.start_time + 1.months).beginning_of_month
```
Comment (sergio-ocon) [inline app/models/showback_bucket.rb]: self.start_time.next_month.all_month

**Pull request merged**